src/02-common-principles.md

```diff
@@ -383,6 +383,200 @@ In particular, if a BIDS dataset contains a `derivatives/` sub-directory,
+[`dataset_description.json` file](./03-modality-agnostic-files.md#dataset-description)
+under the `DatasetLinks` field,
+which is an [object][] of [URIs][], as shown in the examples below.
+
+In brief,
+`dataset_description.DatasetLinks` contains mappings from strings to URIs that point to dataset locations.
+The `<dataset-name>`: `""` (that is, an empty string)
+is a reserved value that may only be used to refer to the root of the current dataset;
+an empty string (`""`) MUST NOT be a key in the `DatasetLinks` object.
+
+In the case where a derivatives dataset is nested under a raw dataset and both have a `dataset_description.json` file,
+the BIDS URIs within the nested derivatives dataset MUST be resolved with respect to `/derivatives/dataset_description.json`,
+and the BIDS URIs for the raw dataset (excluding `/derivatives`) MUST be resolved with respect to `/dataset_description.json`.
+
+However if a derivatives dataset is nested under a raw dataset and only the raw dataset has a `dataset_description.json` file,
+all BIDS URIs MUST be resolved with respect to the `/dataset_description.json` file.
+
+(Assuming that in the two above examples `/` refers to the root of the BIDS raw dataset with a nested derivatives dataset.)
+
+### Refer to a file within a dataset
+
+To link for example to a derivative file that is stored in a nested `derivatives/`
+directory within a dataset:
+`bids:deriv1:/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz`
+
+```json
+{
+    "DatasetLinks": {
+        "deriv1": "file://derivatives/derivative1"
+    }
+
+}
+```
+
+Alternatively, we can also make use of the reserved `<dataset-name>`: `""` (an empty string),
+which always points to the root of the current dataset, to point to the same file using the following syntax:
+`bids::/derivatives/derivative1/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz`
+
+Thus, for this specific case there are two ways to refer to the same file:
+
+1.  `bids:deriv1:/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz`,
+    with `deriv1` being specified within `dataset_description.DatasetLinks`,
+    as shown in the example above.
+
+1.  `bids::/derivatives/derivative1/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz`,
+    with `""` (an empty string) NOT being a key in `dataset_description.DatasetLinks`,
+    because it is a reserved value that always points to the root of the current dataset.
+
+**If *all* BIDS URIs in a dataset *only* use an empty string (`""`) as a `<dataset-name>`,**
+**the `DatasetLinks` metadata MAY not be specified.**
+
+### Refer to a remote dataset
+
+To link for example to a derivative file that is stored in a `derivatives/`
+directory that is stored on a remote server called `mydatahost.com`:
+`bids:deriv2:/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz`
+
+```json
+{
+    "DatasetLinks": {
+        "deriv2": "https://mydatahost.com/derivatives/derivative2"
+    }
+}
+```
+
+This means that the needed files can be obtained by downloading the data
```

Comment:
"downloading the data" -> does that imply downloading the whole dataset, or should it be possible to download only the specified file? See also below.

Reply:
see: b82f7ad

Reply:
>  It now also occurs to me that the DOI often (i.e. in OpenNeuro, Zenodo, and also for the Donders Repository) resolves to a landing page URL, but that the files are not immediately present by extending that URL.

True, I think this point is another argument in favor of "way 1" described in https://github.com/bids-standard/bids-specification/pull/820#discussion_r685942684

Reply:
But this begs the question: what if I want to link to a file that is somewhere on the internet, but _not_ in a BIDS dataset? E.g. if I were to link to `https://github.com/spm/spm12/blob/master/canonical/single_subj_T1.nii?raw=true`. In that case I cannot download the "dataset", but it is also not clear if I should download the SPM12 package and (in the LinkedFile) point to the full path+filename relative to the spm12-master.

The "bids" in `bids:xxx:yyy` indicates that I should resolve the LinkedFile as `dataset_description.DatasetLinks[xxx]` followed by `yyy` relative to the root. It does not indicate that xxx or yyy are in BIDS format.

Reply:
I tried to clarify this by adding the following:

>In the context of BIDS URIs, a dataset with a given `<dataset-name>` does not have to be a BIDS dataset, or a neuroimaging dataset. It may just as well be a (sub-)directory containing one or more arbitrary files.

As for your question:

>  if I were to link to https://github.com/spm/spm12/blob/master/canonical/single_subj_T1.nii?raw=true. In that case I cannot download the "dataset", but it is also not clear if I should download the SPM12 package and (in the LinkedFile) point to the full path+filename relative to the spm12-master.

I added more recommendations (also based on your feedback above) for this. In brief:

Users should first try to make this machine readable. For your example:

- location: `https://github.com/spm/spm12/raw/master/canonical`
- path: `/single_subj_T1.nii`
- full: https://github.com/spm/spm12/raw/master/canonical/single_subj_T1.nii

And if they can't do that (remote host is unreliable/unstable/unfitting), they may link to the smallest possible, **downloadable** subfolder, and then provide a path with respect to that downloaded subfolder.

For your example, this would probably be downloading the GitHub zip of the repo, and providing a path.

if THAT is impossible, they should download the file themselves, and upload it to an easier interface, and then link to that (assuming that licensing issues are being dealt with).

All other cases are hopefully rare enough to never be a true problem.

Tips and Tricks for designing specific links with common hosts (e.g., GitHub) would be content for the starter-kit I suppose. cc @Remi-Gau